src/crimson/os/seastore/cache.h

```diff
@@ -585,34 +626,42 @@ class Cache {
       cached->state = CachedExtent::extent_state_t::INVALID;
       extent_init_func(*ret);
       return read_extent<T>(
-	std::move(ret));
-    } else if (!cached->is_fully_loaded()) {
-      auto ret = TCachedExtentRef<T>(static_cast<T*>(cached.get()));
-      on_cache(*ret);
-      SUBDEBUG(seastore_cache,
-        "{} {}~{} is present without been fully loaded, reading ... -- {}",
-        T::TYPE, offset, length, *ret);
-      auto bp = alloc_cache_buf(length);
-      ret->set_bptr(std::move(bp));
-      return read_extent<T>(
-        std::move(ret));
+	std::move(ret), e_off, e_len);
     } else {
       SUBTRACE(seastore_cache,
           "{} {}~{} is present in cache -- {}",
           T::TYPE, offset, length, *cached);
       auto ret = TCachedExtentRef<T>(static_cast<T*>(cached.get()));
       on_cache(*ret);
       return ret->wait_io(
-      ).then([ret=std::move(ret)]() mutable
+      ).then([ret=std::move(ret), e_off, e_len, this]() mutable
```

Comment:
I think there is another missing piece: after fine-grained-cache, the extent in cache/transaction is possible to be partial loaded/ing now, which means the extent just queried from cache/transaction might be `!is_fully_loaded()` and is waiting for io or not, which means there are more corner cases to handle, for example:
* The partial loading extent might be keeping pending even after `wait_io()` is done, if it is not fully loaded originally.
* The partial loaded extent should not be exposed to the user in all the other scenarios.

Reply:
added a recursive function to resolve this.